server/aesdsocket.c

```diff
@@ -233,148 +479,161 @@ int main(int argc, char **argv)
+            free(new_node);
+            break;
+        }
+        //Insert the node to the linked list
+        new_node->next = head;
+        head = new_node;
 
-        // Print the IP address of the client.
-        struct sockaddr_in *p = (struct sockaddr_in *)&test_addr;
-        syslog(LOG_DEBUG, "Accepted connection from %s", inet_ntop(AF_INET, &p->sin_addr, address_string, sizeof(address_string)));
+        //Check for thread completion.
+        ll_clean(&head, 0);
+    }
 
-        // Receive bytes by polling the client fd
-        while((bytes_rcvd = recv(client_fd, buffer, sizeof(buffer), 0)) > 0)
-        {
-            printf("bytes rcvd = %d\n", bytes_rcvd);
+    /* Clean the linked list and join all the threads. */
+    ll_clean(&head, 1);
 
-            for(int i=0; i < bytes_rcvd; i++)
-            {
-                if(buffer[i] == '\n')
-                {
-                    complete_flag = 1;
-                    actual_bytes_rcvd = i+1;
-                    break;
-                }
-            }
-            memcpy(storage_buffer + (storage_buffer_size_cnt - 1)*INIT_BUF_SIZE, buffer, bytes_rcvd);
+    //cleanup.
+    if(-1 == close(sfd))
+    {
+        perror("close sfd failed: ");
+    }
 
-            if(complete_flag == 1)
-            {
-                bytes_to_write = (storage_buffer_size_cnt - 1)*INIT_BUF_SIZE + actual_bytes_rcvd;
-                break;
-            }
-            else    // reallocated the buffer by multiples of 1024 until complete flag is set
-            {
-                char *temp_ptr = (char *)realloc(storage_buffer, (storage_buffer_size + INIT_BUF_SIZE));
-                if(temp_ptr)
-                {
-                    storage_buffer = temp_ptr;
-                    storage_buffer_size += INIT_BUF_SIZE;
-                    storage_buffer_size_cnt += 1;
-                }
-                else
-                {
-                    printf("No memory available to alloc\n");
-                }
-            }
-        }
-        printf("Storage buffer size = %d\n", bytes_to_write);
+    if (close(fd) == -1)
+    {
+        perror("close fd failed: ");
+    }
 
-        // If transmission is complete,
-        // Copy to local fd and then rewrite/resend it back to client.
-        if(1 == complete_flag)
-        {
-            //Write to file
-            int wr_bytes = write(fd, storage_buffer, bytes_to_write);
-            file_bytes += wr_bytes;
+    //unlink file
+    if (unlink(SOCKET_ADDR) == -1)
+    {
+        perror("unlink failed: ");
+    }
+
+    /* Thread join timer threads */
+    pthread_join(timer_node.thread_id, NULL);
 
-            //Set the fd to start of the file.
-            lseek(fd, 0, SEEK_SET);
+    //Destroy pthread
+    pthread_mutex_destroy(&mutex);
 
-            read_buffer = (char *)malloc(file_bytes);
-            if(!read_buffer)
+    //Close log
+    closelog();
+
+    return 0;
+}
+
+void ll_clean(node_t **head, int is_clear)
```

Comment:
@dash6424 I like the idea to handle the end cleanup differently.  You may consider a bool type here, and some notes about what this flag does.